Update plugin best practices (#668)
napari/docs: docs/plugins/building_a_plugin/best_practices.md

(best-practices)=

> # Best practices

Reviewer: This page would benefit from a new title: Best practices for plugin authors

> # Best practices

Author: ```suggestion
# Best practices for plugin developers
```

(best-practices-no-qt-backend)=

> ## Don't include `napari[all]`, `PySide2`, `PyQt5` or `PyQt6` in your plugin's dependencies.

Reviewer 2: ```suggestion
## Don't include `napari[all]`, `PySide2`, `PyQt5` or `PyQt6` in your plugin's default dependencies.
```
or some similar variation of language. novice's don't quite understand the difference between 'dependencies' and 'optional dependcies' so this might make it more clear

> ## Don't include `napari[all]`, `PySide2`, `PyQt5` or `PyQt6` in your plugin's dependencies.

Reviewer: Maybe `## Avoid including any form of Qt in your plugin's dependencies.

Author: I think I prefer the heading being explicit, but I added your line directly below it.

> *This is important!*

Author: ```suggestion
*This is important! Avoid including any form of Qt in your plugin's dependencies!*
```

Napari supports *both* PyQt and PySide backends for Qt.  It is up to the
end-user to choose which one they want. If they installed napari with `pip
install napari[all]`, then this includes `PyQt5` from PyPI as the default backend.
If they installed via `conda install napari pyqt`, then they'll have `PyQt5`,

> but from conda-forge instead of PyPI. Meanwhile, the napari bundle has PySide2.

Reviewer 2: ```suggestion
but from conda-forge instead of PyPI. Meanwhile, the napari bundle installs with PySide2.
```

> But users are also free to install PyQt6, which is fully supported, or the 

Reviewer 2: ```suggestion
Users are also free to install PyQt6, which is fully supported, or the 
```

  via `pip` (or vice versa) then there *will* be a binary incompatibility between the 
  conda `pyqt` installation, and the `PyQt5` installation from PyPI. *This will very likely
  lead to a broken environment, forcing the user to re-create their entire
  environment and re-install napari*. This is an unfortunate consequence of
  [package naming decisions](https://github.com/ContinuumIO/anaconda-issues/issues/1554),
  and it's not something napari can fix.
- Alternatively, they may end up with some combination of *both* PyQt5, PyQt6, PySide2,
  and PySide6 in their environment: the Qt backend they had installed and the one your
  plugin installed as a dependency. This is will not *always* to break things, but
  it will lead to unexpected and difficult to debug problems. 
- Both of the above cases are most likely to happen with the built-in GUI napari plugin manager,
  which will install your plugin plus the base dependencies. As a result, this frequently 
  occurs with the bundle app. Trying to fix these issues is almost impossible for GUI centric
  users, leaving them the only recourse of re-installing.

> - **Don't import from PyQt5 or PySide2 in your plugin: use `qtpy`.**

Reviewer 2: ```suggestion
## Don't import from PyQt5 or PySide2 in your plugin: use `qtpy`.**
```
Let's separate this out *and* make it more obvious

Author: agreed. I rephrased it a bit.

> - **Don't import from PyQt5 or PySide2 in your plugin: use `qtpy`.**

Author: ```suggestion
## Don't import from any specific Qt backend (e.g. PyQt5, PySide2, etc.) in your plugin: use `qtpy`
```

    If you use `from PyQt5 import QtCore` (or similar) in your plugin, but the
    end-user has chosen to use `PySide2` for their Qt backend — or vice versa —
    then your plugin will fail to import.  Instead use `from qtpy import
    QtCore`.  `qtpy` is a [Qt compatibility layer](https://github.com/spyder-ide/qtpy)
    that will import from whatever backend is installed in the environment.

> ````{tip}

Reviewer 2: move this admonition above the don't import from... section. Will flow better I think

Author: done

If you are vendoring other projects, please add an acknowledgement in your README.
The license details in your project metadata should also include this information!
```

> ## Outdated, npe1 only: Don't import heavy dependencies at the top of your module

Reviewer 2: So, this is outdated with respect to napari's boot up time, but is still good advice for designing libraries.

The default 'python teaching' instructions teach every module to be imported in `__init__.py`, so if a heavy dependency is imported at top level of a module, then *starting the plugin* (but not napari) will be extra slow. 

By continuing to encourage putting heavy imports inside funs/classes then it will reduce overall plugin/library import time. This is especially helpful because we encourage plugins be libraries that may have use outside of napari. 

This is all things I had to learn from experience.

Author: I don't disagree, but think this whole section would need reworking given that the text and code references hook specs. I wasn't feeling quite up-to it, hence marking it out-dated. Maybe you want to take a pass at it in a PR to this PR?
